pennylane/pauli/conversion.py

```diff
@@ -94,52 +110,123 @@ def pauli_decompose(
     + 1.0 * Y(a) @ Y(b)
     + -0.5 * Z(a) @ X(b)
     + -0.5 * Z(a) @ Y(b)
-    """
-    n = int(np.log2(len(H)))
-    N = 2**n
 
-    if wire_order is not None and len(wire_order) != n:
-        raise ValueError(
-            f"number of wires {len(wire_order)} is not compatible with number of qubits {n}"
-        )
+    .. details::
+        :title: Usage Details
+        :href: usage-decompose-operation
 
-    if wire_order is None:
-        wire_order = range(n)
+        For non-square matrices, we need to provide the ``padding=True`` keyword argument:
+
+        >>> A = np.array([[-2, -2 + 1j]])
+        >>> H = qml.pauli_decompose(A, padding=True)
+        >>> print(H)
+          ((-1+0j)) [I0]
+        + ((-1+0.5j)) [X0]
+        + ((-1+0j)) [Z0]
+        + ((-0.5-1j)) [Y0]
+
+        We can also use the method within a differentiable workflow and obtain gradients:
+
+        >>> A = qml.numpy.array([[-2, -2 + 1j]], requires_grad=True)
+        >>> dev = qml.device("default.qubit", wires=1)
+        >>> @qml.qnode(dev)
+        ... def circuit(A):
+        ...    decomp = qml.pauli_decompose(A, padding=True)
+        ...    qml.RX(decomp.coeffs[2], 0)
+        ...    return qml.expval(qml.PauliZ(0))
+        >>> grad_numpy = qml.grad(circuit)(A)
+        tensor([[-2.+0.j, -2.+1.j]], requires_grad=True)
 
-    if H.shape != (N, N):
-        raise ValueError("The matrix should have shape (2**n, 2**n), for any qubit number n>=1")
+    """
+    # Pad with zeros to make the matrix shape equal and a power of two.
+    if padding:
+        shape = qml.math.shape(matrix)
+        num_qubits = int(qml.math.ceil(qml.math.log2(qml.math.max(shape))))
+        if shape[0] != shape[1] or shape[0] != 2**num_qubits:
+            padd_diffs = qml.math.abs(qml.math.array(shape) - 2**num_qubits)
+            padding = (
+                ((0, padd_diffs[0]), (0, padd_diffs[1]))
+                if qml.math.get_interface(matrix) != "torch"
+                else ((padd_diffs[0], 0), (padd_diffs[1], 0))
+            )
+            matrix = qml.math.pad(matrix, padding, mode="constant", constant_values=0)
 
-    if not np.allclose(H, H.conj().T):
-        raise ValueError("The matrix is not Hermitian")
+    shape = qml.math.shape(matrix)
+    if shape[0] != shape[1]:
+        raise ValueError(
+            f"The matrix should be square, got {shape}. Use 'padding=True' for rectangular matrices."
+        )
 
-    obs_lst = []
-    coeffs = []
+    num_qubits = int(qml.math.log2(shape[0]))
+    if shape[0] != 2**num_qubits:
+        raise ValueError(
+            f"Dimension of the matrix should be a power of 2, got {shape}. Use 'padding=True' for these matrices."
+        )
 
```

Comment:
I would prefer not to do this by default for users not to do unintentional shape mismatches while defining their arrays. I guess it would make it difficult for them to debug their code.